### PR TITLE
Add typescript examples for Social sign-in Oauth (react/react-native)

### PR DESCRIPTION
**Pull request by indieisaconcept**
#### Description of changes:
The scope of this change is to provide typescript examples for the [Social sign-in (Oauth)] section of the the Amplify documentation for authentication. Examples have been provided for both react and react-native.

Note whilst I have provided a typescript example for "withOAuth Higher-Order Component (HOC)" I see that `aws-amplify-react-native` used in the implementation appears to be on the path to deprecation as per https://github.com/aws-amplify/amplify-js/issues/2286#issuecomment-1272092888 and looks to have been superseded by `@aws-amplify/ui-react-native`. Is there a revised recommended alternative for `useOauth` when using react-native? If there is I would be happy to raise a subsequent PR to revise the example to reflect it.

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [x] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [x] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 1]
```diff
@@ -53,6 +53,73 @@ import all1 from "/src/fragments/lib/auth/common/social_signin_web_ui/configure_
...
 
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import { useEffect, useState } from "react";
+import { Text, View, Linking, Button } from "react-native";
+import { CognitoHostedUIIdentityProvider } from "@aws-amplify/auth";
+import type { AmplifyUser } from "@aws-amplify/ui";
+import type { HubCapsule } from "@aws-amplify/core";
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:64]:
```suggestion
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 2]
```diff
@@ -53,6 +53,72 @@ import all1 from "/src/fragments/lib/auth/common/social_signin_web_ui/configure_
...
+import awsconfig from "./aws-exports";
+
+Amplify.configure(awsconfig);
+
+export default function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
+  const [customState, setCustomState] = useState<string | null>(null);
+
+  useEffect(() => {
+    const unsubscribe = Hub.listen("auth", ({ payload: { event, data } }: HubCapsule) => {
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:74]:
```suggestion
    const unsubscribe = Hub.listen("auth", ({ payload: { event, data }}) => {
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 3]
```diff
@@ -225,6 +293,101 @@ By default, Amplify will open the Cognito Hosted UI in Safari/Chrome, but you ca
...
+
+```ts
+import { useEffect, useState } from 'react';
+import { Button, Linking, Text, View } from 'react-native';
+import InAppBrowser from 'react-native-inappbrowser-reborn';
+
+import { Amplify, Auth, Hub } from 'aws-amplify';
+import { CognitoHostedUIIdentityProvider } from '@aws-amplify/auth';
+import type { AmplifyUser } from '@aws-amplify/ui';
+import type { HubCapsule } from '@aws-amplify/core';
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:307]:
```suggestion
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 4]
```diff
@@ -225,6 +293,101 @@ By default, Amplify will open the Cognito Hosted UI in Safari/Chrome, but you ca
...
+    ...awsconfig.oauth,
+    urlOpener,
+  },
+});
+
+function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
+
+  useEffect(() => {
+    Hub.listen('auth', ({ payload: { event, data } }: HubCapsule) => {
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:336]:
```suggestion
    Hub.listen('auth', ({ payload: { event, data } }) => {
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 5]
```diff
@@ -322,6 +487,147 @@ By default, Amplify will open the Cognito Hosted UI in Safari/Chrome, but you ca
...
+```ts
+import { useEffect, useState } from "react";
+import { StatusBar } from "expo-status-bar";
+import { StyleSheet, Text, View, Linking, Button } from "react-native";
+import * as WebBrowser from "expo-web-browser";
+
+import { Amplify, Hub, Auth } from "aws-amplify";
+import { CognitoHostedUIIdentityProvider } from "@aws-amplify/auth";
+import type { AmplifyUser } from "@aws-amplify/ui";
+import type { HubCapsule } from "@aws-amplify/core";
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:502]:
```suggestion
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 6]
```diff
@@ -322,6 +487,147 @@ By default, Amplify will open the Cognito Hosted UI in Safari/Chrome, but you ca
...
+  },
+};
+Amplify.configure(updatedConfig);
+
+export default function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
+  const [customState, setCustomState] = useState<string | null>(null);
+
+  useEffect(() => {
+    const unsubscribe = Hub.listen("auth", ({ payload: { event, data } }: HubCapsule) => {
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:545]:
```suggestion
    const unsubscribe = Hub.listen("auth", ({ payload: { event, data }}) => {
```

File: src/fragments/lib/auth/js/social.mdx [hunk 7]
```diff
@@ -53,6 +53,68 @@ import all1 from "/src/fragments/lib/auth/common/social_signin_web_ui/configure_
 
 After configuring the OAuth endpoints (Cognito Hosted UI), you can integrate your App by invoking `Auth.federatedSignIn()` function. Passing `LoginWithAmazon`, `Facebook`, `Google`, or `SignInWithApple` on the `provider` argument (e.g `Auth.federatedSignIn({ provider: 'LoginWithAmazon' })`) will bypass the Hosted UI and federate immediately with the social provider as shown in the below React example. If you are looking to add a custom state, you are able to do so by passing a string (e.g. `Auth.federatedSignIn({ customState: 'xyz' })`) value and listening for the custom state via Hub.
 
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import React, { useEffect, useState } from 'react';
+import { Amplify, Auth, Hub } from 'aws-amplify';
+import type { HubCapsule } from '@aws-amplify/core';
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:62]:
```suggestion
```

File: src/fragments/lib/auth/js/social.mdx [hunk 8]
```diff
@@ -53,6 +53,68 @@ import all1 from "/src/fragments/lib/auth/common/social_signin_web_ui/configure_
...
+import awsConfig from './aws-exports';
+
+Amplify.configure(awsConfig);
+
+function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
+  const [customState, setCustomState] = useState<string | null>(null);
+
+  useEffect(() => {
+    const unsubscribe = Hub.listen("auth", ({ payload: { event, data } }: HubCapsule) => {
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:74]:
```suggestion
    const unsubscribe = Hub.listen("auth", ({ payload: { event, data }}) => {
```

File: src/fragments/lib/auth/js/social.mdx [hunk 9]
```diff
@@ -154,6 +218,98 @@ Amplify.configure(updatedAwsConfig);
 
 ### Full Sample
 
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import { useEffect, useState } from 'react';
+import { Amplify, Auth, Hub } from 'aws-amplify';
+import type { HubCapsule } from '@aws-amplify/core';
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:227]:
```suggestion
```

File: src/fragments/lib/auth/js/social.mdx [hunk 10]
```diff
@@ -154,6 +218,98 @@ Amplify.configure(updatedAwsConfig);
...
+  }
+}
+
+Amplify.configure(updatedAwsConfig);
+
+function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
+
+  useEffect(() => {
+    Hub.listen('auth', ({ payload: { event, data } }: HubCapsule) => {
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:267]:
```suggestion
    Hub.listen('auth', ({ payload: { event, data }}) => {
```

File: src/fragments/lib/auth/js/social.mdx [hunk 11]
```diff
@@ -234,10 +390,30 @@ function App() {
...
 ### Catching Errors
 
 In order to catch errors when using `Auth.federatedSignIn()` you can use the [Hub](/lib/utilities/hub/q/platform/js/) eventing system provided by Amplify to listen for the `signIn_failure` event.
 
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import { Hub } from 'aws-amplify'
+import type { HubCapsule } from '@aws-amplify/core';
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:405]:
```suggestion
```

File: src/fragments/lib/auth/js/social.mdx [hunk 12]
```diff
@@ -234,10 +390,30 @@ function App() {
...
+
+Hub.listen('auth', (data: HubCapsule) => {
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:407]:
```suggestion
Hub.listen('auth', (data) => { 
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 13]
```diff
@@ -53,6 +53,72 @@ import all1 from "/src/fragments/lib/auth/common/social_signin_web_ui/configure_
...
+import { Text, View, Linking, Button } from "react-native";
+import { CognitoHostedUIIdentityProvider } from "@aws-amplify/auth";
+import type { AmplifyUser } from "@aws-amplify/ui";
+import { Amplify, Auth, Hub } from "aws-amplify";
+import awsconfig from "./aws-exports";
+
+Amplify.configure(awsconfig);
+
+export default function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:70]:
```suggestion
  const [user, setUser] = useState(null);
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 14]
```diff
@@ -53,6 +53,72 @@ import all1 from "/src/fragments/lib/auth/common/social_signin_web_ui/configure_
...
 After configuring the OAuth endpoints (Cognito Hosted UI), you can integrate your App by invoking `Auth.federatedSignIn()` function. Passing `LoginWithAmazon`, `Facebook`, `Google`, or `SignInWithApple` on the `provider` argument (e.g `Auth.federatedSignIn({ provider: 'LoginWithAmazon' })`) will bypass the Hosted UI and federate immediately with the social provider as shown in the below React Native example. If you are looking to add a custom state, you are able to do so by passing a string (e.g. `Auth.federatedSignIn({ customState: 'xyz' })`) value and listening for the custom state via Hub.
 
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import { useEffect, useState } from "react";
+import { Text, View, Linking, Button } from "react-native";
+import { CognitoHostedUIIdentityProvider } from "@aws-amplify/auth";
+import type { AmplifyUser } from "@aws-amplify/ui";
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:63]:
```suggestion
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 15]
```diff
@@ -225,6 +293,100 @@ By default, Amplify will open the Cognito Hosted UI in Safari/Chrome, but you ca
...
+<Block name="TypeScript">
+
+```ts
+import { useEffect, useState } from 'react';
+import { Button, Linking, Text, View } from 'react-native';
+import InAppBrowser from 'react-native-inappbrowser-reborn';
+
+import { Amplify, Auth, Hub } from 'aws-amplify';
+import { CognitoHostedUIIdentityProvider } from '@aws-amplify/auth';
+import type { AmplifyUser } from '@aws-amplify/ui';
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:306]:
```suggestion
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 16]
```diff
@@ -225,6 +293,100 @@ By default, Amplify will open the Cognito Hosted UI in Safari/Chrome, but you ca
...
+Amplify.configure({
+  ...awsconfig,
+  oauth: {
+    ...awsconfig.oauth,
+    urlOpener,
+  },
+});
+
+function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:332]:
```suggestion
  const [user, setUser] = useState(null);
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 17]
```diff
@@ -322,6 +486,146 @@ By default, Amplify will open the Cognito Hosted UI in Safari/Chrome, but you ca
...
+
+```ts
+import { useEffect, useState } from "react";
+import { StatusBar } from "expo-status-bar";
+import { StyleSheet, Text, View, Linking, Button } from "react-native";
+import * as WebBrowser from "expo-web-browser";
+
+import { Amplify, Hub, Auth } from "aws-amplify";
+import { CognitoHostedUIIdentityProvider } from "@aws-amplify/auth";
+import type { AmplifyUser } from "@aws-amplify/ui";
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:500]:
```suggestion
```

File: src/fragments/lib/auth/js/react-native-social.mdx [hunk 18]
```diff
@@ -322,6 +486,146 @@ By default, Amplify will open the Cognito Hosted UI in Safari/Chrome, but you ca
...
+    redirectSignOut: isLocalHost
+      ? localRedirectSignOut
+      : productionRedirectSignOut,
+    urlOpener,
+  },
+};
+Amplify.configure(updatedConfig);
+
+export default function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-social.mdx:539]:
```suggestion
  const [user, setUser] = useState(null);
```

File: src/fragments/lib/auth/js/react-native-withoauth.mdx [hunk 19]
```diff
@@ -26,6 +26,77 @@ Although `urlOpener` is left out of this sample for brevity, it is still recomme
...
+<Block name="TypeScript">
+
+```ts
+import React from 'react';
+import { Button, Text, View } from 'react-native';
+import { Amplify, Auth, Hub } from 'aws-amplify';
+import { withOAuth } from "aws-amplify-react-native";
+
+import type { GestureResponderEvent } from 'react-native';
+import type { AmplifyUser } from "@aws-amplify/ui";
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-withoauth.mdx:39]:
```suggestion
```

File: src/fragments/lib/auth/js/react-native-withoauth.mdx [hunk 20]
```diff
@@ -26,6 +26,77 @@ Although `urlOpener` is left out of this sample for brevity, it is still recomme
...
+
+import awsconfig from './aws-exports';
+
+Amplify.configure(awsconfig);
+
+type OnPressEventHandler = (e: GestureResponderEvent) => void;
+
+type AuthProps = {
+  loading?: boolean;
+  oAuthUser?: AmplifyUser;
```
Comment (cwomack) [inline src/fragments/lib/auth/js/react-native-withoauth.mdx:49]:
```suggestion
  oAuthUser?: any;
```

File: src/fragments/lib/auth/js/social.mdx [hunk 21]
```diff
@@ -53,6 +53,67 @@ import all1 from "/src/fragments/lib/auth/common/social_signin_web_ui/configure_
 
 After configuring the OAuth endpoints (Cognito Hosted UI), you can integrate your App by invoking `Auth.federatedSignIn()` function. Passing `LoginWithAmazon`, `Facebook`, `Google`, or `SignInWithApple` on the `provider` argument (e.g `Auth.federatedSignIn({ provider: 'LoginWithAmazon' })`) will bypass the Hosted UI and federate immediately with the social provider as shown in the below React example. If you are looking to add a custom state, you are able to do so by passing a string (e.g. `Auth.federatedSignIn({ customState: 'xyz' })`) value and listening for the custom state via Hub.
 
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import React, { useEffect, useState } from 'react';
+import { Amplify, Auth, Hub } from 'aws-amplify';
+import type { AmplifyUser } from '@aws-amplify/ui';
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:62]:
```suggestion
```

File: src/fragments/lib/auth/js/social.mdx [hunk 22]
```diff
@@ -53,6 +53,67 @@ import all1 from "/src/fragments/lib/auth/common/social_signin_web_ui/configure_
...
+import { CognitoHostedUIIdentityProvider } from '@aws-amplify/auth';
+import awsConfig from './aws-exports';
+
+Amplify.configure(awsConfig);
+
+function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:69]:
```suggestion
  const [user, setUser] = useState(null);
```

File: src/fragments/lib/auth/js/social.mdx [hunk 23]
```diff
@@ -154,6 +217,97 @@ Amplify.configure(updatedAwsConfig);
 
 ### Full Sample
 
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import { useEffect, useState } from 'react';
+import { Amplify, Auth, Hub } from 'aws-amplify';
+import type { AmplifyUser } from "@aws-amplify/ui";
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:226]:
```suggestion
```

File: src/fragments/lib/auth/js/social.mdx [hunk 24]
```diff
@@ -154,6 +217,97 @@ Amplify.configure(updatedAwsConfig);
...
+    ...awsConfig.oauth,
+    redirectSignIn: isLocalhost ? localRedirectSignIn : productionRedirectSignIn,
+    redirectSignOut: isLocalhost ? localRedirectSignOut : productionRedirectSignOut,
+  }
+}
+
+Amplify.configure(updatedAwsConfig);
+
+function App() {
+  const [user, setUser] = useState<AmplifyUser | null>(null);
```
Comment (cwomack) [inline src/fragments/lib/auth/js/social.mdx:262]:
```suggestion
  const [user, setUser] = useState(null);
```

**Review by israx — APPROVED** [review]
Hey @indieisaconcept . Thanks so much for putting all of these together. LGTM

**Review by abdallahshaban557 — APPROVED** [review]
LGTM

**Review by cwomack — APPROVED** [review]
LGTM

**Review by katiegoines — APPROVED** [review]

**Pull request merged**